lib/core/proxy.c

```diff
@@ -283,7 +284,17 @@ static void do_send(struct rp_generator_t *self)
         veccnt = 1;
         is_eos = 0;
     }
-    h2o_send(self->src_req, vecs, veccnt, is_eos);
+    h2o_send_state_t ststate;
+    if (self->had_body_error) {
+        ststate = H2O_SEND_STATE_ERROR;
+    } else {
+        if (is_eos) {
+            ststate = H2O_SEND_STATE_FINAL;
+        } else {
+            ststate = H2O_SEND_STATE_IN_PROGRESS;
+        }
+    }
+    h2o_send(self->src_req, vecs, veccnt, ststate);
```

Comment:
It might be easier to change `int is_eos;` to `h2o_send_state_t ststate;`, and assign values directly to the state as we did for `is_eos` prior to this PR (see https://github.com/h2o/h2o/blob/b1faa94/lib/core/proxy.c#L277-L284).